### ISSUE TITLE
Dataset preparation

### ISSUE THREAD

**Issue by makhovds** [Zhaozixiang1228/GDSR-DCTNet#14]
Good day!
Thank you for the paper with such an interesting idea of application of DCT. I want to reproduce you result on NYU v2 dataset.
I am extracting square patches 256x256 with stride 100 and use nearest upsampling/downsampling to achieve LR patch of the same shape. Using this code NYU_Train_imgsize_256_scale_4.h5 file weights 5.98 GB but not 10+ as you mention. I tried training on it and the RMSE at the testing stage is awful - 300.89. So it seems that the reproduction fails. Could you please share clear instruction of how to prepare the dataset or share it directly?

**Note**
I tried to create the dataset with the following code:
```python
from scipy.io import loadmat
import numpy as np
import h5py
from resize_right.resize_right import resize
from resize_right.interp_methods import cubic, box
from torch.nn.functional import upsample_nearest
import torch
from patches import extract_patches_2d
from shutil import rmtree
from os.path import isfile, isdir
from os import remove, makedirs
import matplotlib.pyplot as plt
from skimage.io import imsave

NYU_ROOT_PATH = "/media/denis_makhov/c192784f-e992-4c1e-9acb-711282eac532/denis/datasets/nyuv2/nyu_depth_v2_labeled.mat"
PATCH_SIZE = (256, 256)
STEP = (100, 100)
# RESIZE_METHOD = 'bicubic'
SCALE = 4

TRAIN_IDXS = [0, 1000]
VAL_IDXS = [0, 20]
TEST_IDXS = [1000, 1449]

# TRAIN_IDXS = [0, 10]
# VAL_IDXS = [0, 2]
# TEST_IDXS = [1445, 1449]

train_name = './data/NYU_Train_imgsize_' + str(PATCH_SIZE[0]) + '_scale_' + str(SCALE) + '.h5'
# test_name = './data/NYU_Test_imgsize_' + str(PATCH_SIZE[0]) + '_scale_' + str(SCALE) + '.h5'
val_name = r'./data/NYU_Validation_AfterProcessing_'+str(SCALE)+'X'
test_name = r'./DatasetsAfterProcessing/NYU_Test_AfterProcessing_'+str(SCALE)+'X'

def process_nyuv2():

    # Read .mat file and parse images and depth
    f = h5py.File(NYU_ROOT_PATH, 'r')
    # print(f.keys())
    # test = f['rawRgbFilenames']
    # st = test[0, 0]
    # obj = f[st]
    # str1 = ''.join(chr(i) for i in obj[:, 0])
    # print( str1 )
    # exit()
    images = f['images']
    depths = f['depths']

    train_images = torch.tensor(images[TRAIN_IDXS[0] : TRAIN_IDXS[1]]).cpu()
    val_images = torch.tensor(images[VAL_IDXS[0] : VAL_IDXS[1]]).cpu()
    test_images = torch.tensor(images[TEST_IDXS[0] : TEST_IDXS[1]]).cpu()
    train_depths_hr = torch.tensor(depths[TRAIN_IDXS[0] : TRAIN_IDXS[1]]).unsqueeze(1).cpu()
    val_depths_hr = torch.tensor(depths[VAL_IDXS[0] : VAL_IDXS[1]]).unsqueeze(1).cpu()
    test_depths_hr = torch.tensor(depths[TEST_IDXS[0] : TEST_IDXS[1]]).unsqueeze(1).cpu()

    del images, depths

    if isfile(train_name):
        remove(train_name)

    for idx, (train_image, train_depth_hr) in enumerate(zip(train_images, train_depths_hr)):
        train_image = extract_patches_2d(img = train_image.unsqueeze(0), patch_shape=PATCH_SIZE, step=STEP, batch_first=True).reshape(-1, 3, PATCH_SIZE[0], PATCH_SIZE[1])
        train_depth_hr = extract_patches_2d(img = train_depth_hr.unsqueeze(0), patch_shape=PATCH_SIZE, step=STEP, batch_first=True).reshape(-1, 1, PATCH_SIZE[0], PATCH_SIZE[1])

        # Downsample depths to RESIZE_SIZE to create an input for the network
        train_depth_lr = upsample_nearest(train_depth_hr, scale_factor=1 / SCALE)
        train_depth_lr = upsample_nearest(train_depth_lr, scale_factor=SCALE)
    
        with h5py.File(train_name, "a") as file:
            if idx != 0:
                file["RGB"].resize((file["RGB"].shape[0] + train_image.shape[0]), axis = 0)
                file["RGB"][-train_image.shape[0]:] = train_image
                file["LRDepth"].resize((file["LRDepth"].shape[0] + train_depth_lr.shape[0]), axis = 0)
                file["LRDepth"][-train_depth_lr.shape[0]:] = train_depth_lr
                file["HRDepth"].resize((file["HRDepth"].shape[0] + train_depth_hr.shape[0]), axis = 0)
                file["HRDepth"][-train_depth_hr.shape[0]:] = train_depth_hr
            else:
                file.create_dataset('RGB', data=train_image, compression="gzip", chunks=True, maxshape=(None, None, None, None))
                file.create_dataset('LRDepth', data=train_depth_lr, compression="gzip", chunks=True, maxshape=(None, None, None, None))
                file.create_dataset('HRDepth', data=train_depth_hr, compression="gzip", chunks=True, maxshape=(None, None, None, None))

    # if isfile(test_name):
    #     remove(test_name)
    # for idx, (test_image, test_depth_hr) in enumerate(zip(test_images, test_depths_hr)):
    #     test_image = extract_patches_2d(img = test_image.unsqueeze(0), patch_shape=PATCH_SIZE, step=STEP, batch_first=True).reshape(-1, 3, PATCH_SIZE[0], PATCH_SIZE[1])
    #     test_depth_hr = extract_patches_2d(img = test_depth_hr.unsqueeze(0), patch_shape=PATCH_SIZE, step=STEP, batch_first=True).reshape(-1, 1, PATCH_SIZE[0], PATCH_SIZE[1])

    #     # Downsample depths to RESIZE_SIZE to create an input for the network
    #     test_depth_lr = upsample_nearest(test_depth_hr, scale_factor=1 / SCALE)
    #     test_depth_lr = upsample_nearest(test_depth_lr, scale_factor=SCALE)
    
    #     with h5py.File(test_name, "a") as file:
    #         if idx != 0:
    #             file["RGB"].resize((file["RGB"].shape[0] + test_image.shape[0]), axis = 0)
    #             file["RGB"][-test_image.shape[0]:] = test_image
    #             file["LRDepth"].resize((file["LRDepth"].shape[0] + test_depth_lr.shape[0]), axis = 0)
    #             file["LRDepth"][-test_depth_lr.shape[0]:] = test_depth_lr
    #             file["HRDepth"].resize((file["HRDepth"].shape[0] + test_depth_hr.shape[0]), axis = 0)
    #             file["HRDepth"][-test_depth_hr.shape[0]:] = test_depth_hr
    #         else:
    #             file.create_dataset('RGB', data=test_image, compression="gzip", chunks=True, maxshape=(None, None, None, None))
    #             file.create_dataset('LRDepth', data=test_depth_lr, compression="gzip", chunks=True, maxshape=(None, None, None, None))
    #             file.create_dataset('HRDepth', data=test_depth_hr, compression="gzip", chunks=True, maxshape=(None, None, None, None))


    ################## VAL PART PREPARATION #################
    if isdir(val_name):
        rmtree(val_name)

    makedirs(val_name, exist_ok=True)
    makedirs(val_name + '/RGB', exist_ok=True)
    makedirs(val_name + '/DepthHR', exist_ok=True)
    makedirs(val_name + '/DepthLrUp', exist_ok=True)

    for idx, (val_image, val_depth_hr) in enumerate(zip(val_images, val_depths_hr)):
        # Downsample depths to RESIZE_SIZE to create an input for the network
        test_depth_lr = upsample_nearest(val_depth_hr, scale_factor=1 / SCALE)
        test_depth_lr = upsample_nearest(test_depth_lr, scale_factor=SCALE)

        np.save(val_name + '/RGB/' + f"{VAL_IDXS[0] + idx}.npy", val_image)
        np.save(val_name + '/DepthHR/' + f"{VAL_IDXS[0] + idx}.npy", val_depth_hr)
        np.save(val_name + '/DepthLrUp/' + f"{VAL_IDXS[0] + idx}.npy", test_depth_lr)


    ################## TEST PART PREPARATION ################
    if isdir(test_name):
        rmtree(test_name)

    makedirs(test_name, exist_ok=True)
    makedirs(test_name + '/RGB', exist_ok=True)
    makedirs(test_name + '/DepthHR', exist_ok=True)
    makedirs(test_name + '/DepthLrUp', exist_ok=True)

    for idx, (test_image, test_depth_hr) in enumerate(zip(test_images, test_depths_hr)):
        # Downsample depths to RESIZE_SIZE to create an input for the network
        test_depth_lr = upsample_nearest(test_depth_hr, scale_factor=1 / SCALE)
        test_depth_lr = upsample_nearest(test_depth_lr, scale_factor=SCALE)

        np.save(test_name + '/RGB/' + f"{TEST_IDXS[0] + idx}.npy", test_image)
        np.save(test_name + '/DepthHR/' + f"{TEST_IDXS[0] + idx}.npy", test_depth_hr)
        np.save(test_name + '/DepthLrUp/' + f"{TEST_IDXS[0] + idx}.npy", test_depth_lr)


def check_dataset():
    f_train = h5py.File(train_name, 'r')
    f_test = h5py.File(test_name, 'r')


    for f, name in zip([f_train, f_test], ['train', 'test']):
        fig, axs = plt.subplots(1, 3)
        axs[0].imshow(f['RGB'][0].transpose((1, 2, 0)))
        axs[1].imshow(f['HRDepth'][0].squeeze())
        axs[2].imshow(f['LRDepth'][0].squeeze())
        fig.savefig(f"check_dataset_{name}.png")
        plt.close()


if __name__ == "__main__":
    process_nyuv2()

    # check_dataset()
```

I extract patches using the following script:
```python
import torch
import torch.nn as nn
import numpy as np

def extract_patches_2d(img,patch_shape,step=[1.0,1.0],batch_first=False):
    patch_H, patch_W = patch_shape[0], patch_shape[1]
    if(img.size(2)<patch_H):
        num_padded_H_Top = (patch_H - img.size(2))//2
        num_padded_H_Bottom = patch_H - img.size(2) - num_padded_H_Top
        padding_H = nn.ConstantPad2d((0,0,num_padded_H_Top,num_padded_H_Bottom),0)
        img = padding_H(img)
    if(img.size(3)<patch_W):
        num_padded_W_Left = (patch_W - img.size(3))//2
        num_padded_W_Right = patch_W - img.size(3) - num_padded_W_Left
        padding_W = nn.ConstantPad2d((num_padded_W_Left,num_padded_W_Right,0,0),0)
        img = padding_W(img)
    step_int = [0,0]
    step_int[0] = int(patch_H*step[0]) if(isinstance(step[0], float)) else step[0]
    step_int[1] = int(patch_W*step[1]) if(isinstance(step[1], float)) else step[1]
    patches_fold_H = img.unfold(2, patch_H, step_int[0])
    if((img.size(2) - patch_H) % step_int[0] != 0):
        patches_fold_H = torch.cat((patches_fold_H,img[:,:,-patch_H:,].permute(0,1,3,2).unsqueeze(2)),dim=2)
    patches_fold_HW = patches_fold_H.unfold(3, patch_W, step_int[1])   
    if((img.size(3) - patch_W) % step_int[1] != 0):
        patches_fold_HW = torch.cat((patches_fold_HW,patches_fold_H[:,:,:,-patch_W:,:].permute(0,1,2,4,3).unsqueeze(3)),dim=3)
    patches = patches_fold_HW.permute(2,3,0,1,4,5)
    patches = patches.reshape(-1,img.size(0),img.size(1),patch_H,patch_W)
    if(batch_first):
        patches = patches.permute(1,0,2,3,4)
    return patches

def reconstruct_from_patches_2d(patches,img_shape,step=[1.0,1.0],batch_first=False):
    if(batch_first):
        patches = patches.permute(1,0,2,3,4)
    patch_H, patch_W = patches.size(3), patches.size(4)
    img_size = (patches.size(1), patches.size(2),max(img_shape[0], patch_H), max(img_shape[1], patch_W))
    step_int = [0,0]
    step_int[0] = int(patch_H*step[0]) if(isinstance(step[0], float)) else step[0]
    step_int[1] = int(patch_W*step[1]) if(isinstance(step[1], float)) else step[1]
    nrow, ncol = 1 + (img_size[-2] - patch_H)//step_int[0], 1 + (img_size[-1] - patch_W)//step_int[1]
    r_nrow = nrow + 1 if((img_size[2] - patch_H) % step_int[0] != 0) else nrow
    r_ncol = ncol + 1 if((img_size[3] - patch_W) % step_int[1] != 0) else ncol
    patches = patches.reshape(r_nrow,r_ncol,img_size[0],img_size[1],patch_H,patch_W)
    img = torch.zeros(img_size, device = patches.device)
    overlap_counter = torch.zeros(img_size, device = patches.device)
    for i in range(nrow):
        for j in range(ncol):
            img[:,:,i*step_int[0]:i*step_int[0]+patch_H,j*step_int[1]:j*step_int[1]+patch_W] += patches[i,j,]
            overlap_counter[:,:,i*step_int[0]:i*step_int[0]+patch_H,j*step_int[1]:j*step_int[1]+patch_W] += 1
    if((img_size[2] - patch_H) % step_int[0] != 0):
        for j in range(ncol):
            img[:,:,-patch_H:,j*step_int[1]:j*step_int[1]+patch_W] += patches[-1,j,]
            overlap_counter[:,:,-patch_H:,j*step_int[1]:j*step_int[1]+patch_W] += 1
    if((img_size[3] - patch_W) % step_int[1] != 0):
        for i in range(nrow):
            img[:,:,i*step_int[0]:i*step_int[0]+patch_H,-patch_W:] += patches[i,-1,]
            overlap_counter[:,:,i*step_int[0]:i*step_int[0]+patch_H,-patch_W:] += 1
    if((img_size[2] - patch_H) % step_int[0] != 0 and (img_size[3] - patch_W) % step_int[1] != 0):
        img[:,:,-patch_H:,-patch_W:] += patches[-1,-1,]
        overlap_counter[:,:,-patch_H:,-patch_W:] += 1
    img /= overlap_counter
    if(img_shape[0]<patch_H):
        num_padded_H_Top = (patch_H - img_shape[0])//2
        num_padded_H_Bottom = patch_H - img_shape[0] - num_padded_H_Top
        img = img[:,:,num_padded_H_Top:-num_padded_H_Bottom,]
    if(img_shape[1]<patch_W):
        num_padded_W_Left = (patch_W - img_shape[1])//2
        num_padded_W_Right = patch_W - img_shape[1] - num_padded_W_Left
        img = img[:,:,:,num_padded_W_Left:-num_padded_W_Right]
    return img

def stitch_v1(patches: torch.Tensor, overlap: tuple):
    """
    Stitch patches that are stored in tensor of shape NHxNWxHxW,
    where NH and NW means number of patches per axis H and W, correspondingly
    """
    nh, nw, h, w = patches.shape
    oh, ow = overlap

    # Calculate output shape of the stitched image
    out_h = h * nh - oh * (nh - 1)
    out_w = w * nw - ow * (nw - 1)

    out = torch.zeros(out_h, out_w)
    for i in range(nh):
        for j in range(nw):
            patch = patches[i, j]
            
            # Generate mask with exponentially decaying weight in overlap regions
            mask = torch.ones_like(patch)
            w = np.linspace(overlap)
            if j == 0: # apply only on right
                mask[:, -overlap:] *= w ** 2 - 2 * w + 1
'''
